KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+    /**
+     * @dev resets all storage values to empty objects except targetBlockNumber
+     */
+    function reset() public onlyOwner {
+        //reset cannot be called at Finalized status or after target block.number
+        require(
+            ((status != Status.Finalized) &&
+                (block.number < rebalanceBlockNumber)),
+            "Contract is finalized, cannot reset values"
+        );
+
+        //`delete` keyword is used to set a storage variable or a dynamic array to its default value.
+        delete retirees;
+        delete newbies;
+        delete memo;
+        status = Status.Initialized;
+    }
+
+    //Getters
+    /**
+     * @dev to get retired details by retiredAddress
+     * @param _retiredAddress is the address of the retired
+     */
+    function getRetired(
+        address _retiredAddress
+    ) public view returns (address, address[] memory) {
+        require(retiredExists(_retiredAddress), "Retired does not exist");
+        uint256 index = getRetiredIndex(_retiredAddress);
+        Retired memory retired = retirees[index];
+        return (retired.retired, retired.approvers);
+    }
+
+    /**
+     * @dev check whether retiredAddress is registered
+     * @param _retiredAddress is the address of the retired
+     */
+    function retiredExists(address _retiredAddress) public view returns (bool) {
+        require(_retiredAddress != address(0), "Invalid address");
+        for (uint8 i = 0; i < retirees.length; i++) {
+            if (retirees[i].retired == _retiredAddress) {
+                return true;
+            }
+        }
+    }
+
+    /**
+     * @dev get index of the retired in the retirees array
+     * @param _retiredAddress is the address of the retired
+     */
+    function getRetiredIndex(
+        address _retiredAddress
+    ) public view returns (uint256) {
+        for (uint256 i = 0; i < retirees.length; i++) {
+            if (retirees[i].retired == _retiredAddress) {
+                return i;
+            }
+        }
+        revert("Retired does not exist");
+    }
+
+    /**
+     * @dev to calculate the sum of retirees balances
+     * @return retireesBalance the sum of balances of retireds
+     */
+    function sumOfRetiredBalance() public view returns (uint256 retireesBalance) {
+        for (uint8 i = 0; i < retirees.length; i++) {
+            address retiredAddress = retirees[i].retired;
+            retireesBalance += retiredAddress.balance;
+        }
+        return retireesBalance;
+    }
+
+    /**
+     * @dev to get newbie details by newbieAddress
+     * @param _newbieAddress is the address of the newbie
+     * @return newbie is the address of the newbie
+     * @return amount is the fund allocated to the newbie
+
+     */
+    function getNewbie(
+        address _newbieAddress
+    ) public view returns (address, uint256) {
+        require(newbieExists(_newbieAddress), "Newbie does not exist");
+        uint256 index = getNewbieIndex(_newbieAddress);
```

Comment:
Why do we need to have same iterations in the two functions?
If we get wrong index, we are able to know that newbie does not exist.

Reply:
Removed unnecessary checks.